// vscode-vue-languageservice/out/parsers/scriptRanges.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseScriptRanges = void 0;
const scriptSetupRanges_1 = require("./scriptSetupRanges");
function parseScriptRanges(ts, content, lang, hasScriptSetup, withComponentOption = false, withNode = false) {
    let exportDefault;
    const sourceFile = ts.createSourceFile('foo.' + lang, content, ts.ScriptTarget.Latest);
    const bindings = hasScriptSetup ? scriptSetupRanges_1.parseBindingRanges(ts, sourceFile) : [];
    sourceFile.forEachChild(node => {
        if (ts.isExportAssignment(node)) {
            let obj;
            if (ts.isObjectLiteralExpression(node.expression)) {
                obj = node.expression;
            }
            else if (ts.isCallExpression(node.expression) && node.expression.arguments.length) {
                const arg0 = node.expression.arguments[0];
                if (ts.isObjectLiteralExpression(arg0)) {
                    obj = arg0;
                }
            }
            if (obj) {
                let componentsOptionNode;
                if (withComponentOption) {
                    obj.forEachChild(node => {
                        if (ts.isPropertyAssignment(node) && ts.isIdentifier(node.name)) {
                            if (node.name.escapedText === 'components' && ts.isObjectLiteralExpression(node.initializer)) {
                                componentsOptionNode = node.initializer;
                            }
                        }
                    });
                }
                exportDefault = {
                    ..._getStartEnd(node),
                    expression: _getStartEnd(node.expression),
                    args: _getStartEnd(obj),
                    argsNode: withNode ? obj : undefined,
                    componentsOption: componentsOptionNode ? _getStartEnd(componentsOptionNode) : undefined,
                    componentsOptionNode: withNode ? componentsOptionNode : undefined,
                };
            }
        }
    });
    return {
        sourceFile,
        exportDefault,
        bindings,
    };
    function _getStartEnd(node) {
        return scriptSetupRanges_1.getStartEnd(node, sourceFile);
    }
}
exports.parseScriptRanges = parseScriptRanges;
