// vscode-vue-languageservice/out/use/useSfcScriptGen.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSfcScriptGen = void 0;
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
const shared = require("@volar/shared");
const reactivity_1 = require("@vue/reactivity");
const sourceMaps_1 = require("../utils/sourceMaps");
const scriptRanges_1 = require("../parsers/scriptRanges");
const scriptSetupRanges_1 = require("../parsers/scriptSetupRanges");
const script_1 = require("../generators/script");
const script_suggestion_1 = require("../generators/script_suggestion");
const templateGen = require("../generators/template_scriptSetup");
function useSfcScriptGen(lsType, ts, vueDoc, script, scriptSetup, html) {
    let version = 0;
    const uri = vueDoc.value.uri;
    const scriptRanges = reactivity_1.computed(() => script.value
        ? scriptRanges_1.parseScriptRanges(ts, script.value.content, script.value.lang, !!scriptSetup.value)
        : undefined);
    const scriptSetupRanges = reactivity_1.computed(() => scriptSetup.value
        ? scriptSetupRanges_1.parseScriptSetupRanges(ts, scriptSetup.value.content, scriptSetup.value.lang)
        : undefined);
    const codeGen = reactivity_1.computed(() => script_1.generate(lsType, uri, script.value, scriptSetup.value, scriptRanges.value, scriptSetupRanges.value));
    const htmlGen = reactivity_1.computed(() => {
        if (html.value) {
            return templateGen.generate(html.value);
        }
    });
    const suggestionCodeGen = reactivity_1.computed(() => script_suggestion_1.generate(script.value, scriptSetup.value, scriptRanges.value, scriptSetupRanges.value, htmlGen.value));
    const lang = reactivity_1.computed(() => {
        return !script.value && !scriptSetup.value ? 'ts'
            : scriptSetup.value && scriptSetup.value.lang !== 'js' ? shared.getValidScriptSyntax(scriptSetup.value.lang)
                : script.value && script.value.lang !== 'js' ? shared.getValidScriptSyntax(script.value.lang)
                    : 'js';
    });
    const textDocument = reactivity_1.computed(() => {
        return vscode_languageserver_textdocument_1.TextDocument.create(lsType === 'template' ? `${uri}.__VLS_script.${lang.value}` : `${uri}.${lang.value}`, shared.syntaxToLanguageId(lang.value), version++, codeGen.value.getText());
    });
    const textDocumentForSuggestion = reactivity_1.computed(() => {
        if (!suggestionCodeGen.value)
            return;
        return vscode_languageserver_textdocument_1.TextDocument.create(`${uri}.__VLS_script.suggestion.${lang.value}`, shared.syntaxToLanguageId(lang.value), version++, suggestionCodeGen.value.getText());
    });
    const sourceMap = reactivity_1.computed(() => {
        var _a;
        const sourceMap = new sourceMaps_1.TsSourceMap(vueDoc.value, textDocument.value, lsType, false, {
            foldingRanges: false,
            formatting: false,
            documentSymbol: true,
            codeActions: true,
            organizeImports: !((_a = script.value) === null || _a === void 0 ? void 0 : _a.src) && !scriptSetup.value,
        }, codeGen.value.getMappings(parseMappingSourceRange));
        return sourceMap;
    });
    const sourceMapForSuggestion = reactivity_1.computed(() => {
        if (!suggestionCodeGen.value)
            return;
        if (!textDocumentForSuggestion.value)
            return;
        const sourceMap = new sourceMaps_1.TsSourceMap(vueDoc.value, textDocumentForSuggestion.value, lsType, false, {
            foldingRanges: false,
            formatting: false,
            documentSymbol: false,
            codeActions: true,
            organizeImports: true,
        }, suggestionCodeGen.value.getMappings(parseMappingSourceRange));
        return sourceMap;
    });
    const teleportSourceMap = reactivity_1.computed(() => {
        const doc = textDocument.value;
        const sourceMap = new sourceMaps_1.TeleportSourceMap(doc, false);
        for (const teleport of codeGen.value.teleports) {
            sourceMap.add(teleport);
        }
        return sourceMap;
    });
    return {
        lang,
        scriptSetupRanges,
        textDocument,
        textDocumentForSuggestion,
        sourceMap,
        sourceMapForSuggestion,
        teleportSourceMap,
    };
    function parseMappingSourceRange(data, sourceRange) {
        var _a;
        if (data.vueTag === 'scriptSrc' && ((_a = script.value) === null || _a === void 0 ? void 0 : _a.src)) {
            const vueStart = vueDoc.value.getText().substring(0, script.value.loc.start).lastIndexOf(script.value.src);
            const vueEnd = vueStart + script.value.src.length;
            return {
                start: vueStart - 1,
                end: vueEnd + 1,
            };
        }
        else if (data.vueTag === 'script' && script.value) {
            return {
                start: script.value.loc.start + sourceRange.start,
                end: script.value.loc.start + sourceRange.end,
            };
        }
        else if (data.vueTag === 'scriptSetup' && scriptSetup.value) {
            return {
                start: scriptSetup.value.loc.start + sourceRange.start,
                end: scriptSetup.value.loc.start + sourceRange.end,
            };
        }
        else {
            return sourceRange;
        }
    }
}
exports.useSfcScriptGen = useSfcScriptGen;
