// @volar/shared/out/requests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoveAllRefSugars = exports.DetectDocumentNameCasesRequest = exports.RestartServerNotification = exports.WriteVirtualFilesRequest = exports.VerifyAllScriptsRequest = exports.GetRefCompleteEditsRequest = exports.GetTagCloseEditsRequest = exports.D3Request = exports.PingRequest = exports.GetDocumentSelectionRequest = exports.GetDocumentNameCasesRequest = exports.ShowReferencesNotification = exports.GetDocumentVersionRequest = exports.GetDocumentContentRequest = void 0;
const rpc = require("vscode-jsonrpc");
/**
 * Client Requests
 */
var GetDocumentContentRequest;
(function (GetDocumentContentRequest) {
    GetDocumentContentRequest.type = new rpc.RequestType('vscode/content');
})(GetDocumentContentRequest = exports.GetDocumentContentRequest || (exports.GetDocumentContentRequest = {}));
var GetDocumentVersionRequest;
(function (GetDocumentVersionRequest) {
    GetDocumentVersionRequest.type = new rpc.RequestType('vue/docUpdated');
})(GetDocumentVersionRequest = exports.GetDocumentVersionRequest || (exports.GetDocumentVersionRequest = {}));
var ShowReferencesNotification;
(function (ShowReferencesNotification) {
    ShowReferencesNotification.type = new rpc.NotificationType('vue.findReferences');
})(ShowReferencesNotification = exports.ShowReferencesNotification || (exports.ShowReferencesNotification = {}));
var GetDocumentNameCasesRequest;
(function (GetDocumentNameCasesRequest) {
    GetDocumentNameCasesRequest.type = new rpc.RequestType('volar/getAttrNameCaseClient');
})(GetDocumentNameCasesRequest = exports.GetDocumentNameCasesRequest || (exports.GetDocumentNameCasesRequest = {}));
var GetDocumentSelectionRequest;
(function (GetDocumentSelectionRequest) {
    GetDocumentSelectionRequest.type = new rpc.RequestType0('vue/activeSelection');
})(GetDocumentSelectionRequest = exports.GetDocumentSelectionRequest || (exports.GetDocumentSelectionRequest = {}));
/**
 * Server Requests
 */
var PingRequest;
(function (PingRequest) {
    PingRequest.type = new rpc.RequestType0('volar/ping');
})(PingRequest = exports.PingRequest || (exports.PingRequest = {}));
var D3Request;
(function (D3Request) {
    D3Request.type = new rpc.RequestType('volar/d3');
})(D3Request = exports.D3Request || (exports.D3Request = {}));
var GetTagCloseEditsRequest;
(function (GetTagCloseEditsRequest) {
    GetTagCloseEditsRequest.type = new rpc.RequestType('html/tag');
})(GetTagCloseEditsRequest = exports.GetTagCloseEditsRequest || (exports.GetTagCloseEditsRequest = {}));
var GetRefCompleteEditsRequest;
(function (GetRefCompleteEditsRequest) {
    GetRefCompleteEditsRequest.type = new rpc.RequestType('volar/ref');
})(GetRefCompleteEditsRequest = exports.GetRefCompleteEditsRequest || (exports.GetRefCompleteEditsRequest = {}));
var VerifyAllScriptsRequest;
(function (VerifyAllScriptsRequest) {
    VerifyAllScriptsRequest.type = new rpc.RequestType('volar.action.verifyAllScripts');
})(VerifyAllScriptsRequest = exports.VerifyAllScriptsRequest || (exports.VerifyAllScriptsRequest = {}));
var WriteVirtualFilesRequest;
(function (WriteVirtualFilesRequest) {
    WriteVirtualFilesRequest.type = new rpc.RequestType('volar.action.writeVirtualFiles');
})(WriteVirtualFilesRequest = exports.WriteVirtualFilesRequest || (exports.WriteVirtualFilesRequest = {}));
var RestartServerNotification;
(function (RestartServerNotification) {
    RestartServerNotification.type = new rpc.NotificationType('volar.action.restartServer');
})(RestartServerNotification = exports.RestartServerNotification || (exports.RestartServerNotification = {}));
var DetectDocumentNameCasesRequest;
(function (DetectDocumentNameCasesRequest) {
    DetectDocumentNameCasesRequest.type = new rpc.RequestType('volar/getTagNameCaseServer');
})(DetectDocumentNameCasesRequest = exports.DetectDocumentNameCasesRequest || (exports.DetectDocumentNameCasesRequest = {}));
var RemoveAllRefSugars;
(function (RemoveAllRefSugars) {
    RemoveAllRefSugars.type = new rpc.NotificationType('volar/removeAllRefSugars');
})(RemoveAllRefSugars = exports.RemoveAllRefSugars || (exports.RemoveAllRefSugars = {}));
